### PR TITLE
Feature/gh 4 package vue and api

### PR DESCRIPTION
**Pull request by mattgen88**
#4 Packages up app and api.
This PR will enable the API to serve up compiled frontend assets.
This PR also cleans up the infrastructure, enabling easier development.
This should also fix the CORS issues, now that I had time to think post-hackathon

### CODE REVIEW

File: .env.dist [hunk 1]
```diff
@@ -0,0 +1 @@
+VUE_APP_API_ORIGIN=http://api.invalid/api
```
Comment (mattgen88) [inline .env.dist]: New environment file. This is used by the Vue app

File: Dockerfile [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
+RUN npm i
+RUN npm run build
+
+FROM alpine as runtime
+RUN apk update && apk add --no-cache ca-certificates
+
 WORKDIR /app
+COPY --from=gobuilder /build/main /app/main
+COPY --from=nodebuilder /build/dist /app/dist
 CMD ["./main"]
```
Comment (mattgen88) [inline Dockerfile]: Mult-stage docker file, builds the API, then the frontend, and brings them together in a runtime environment.

File: Makefile [hunk 3]
```diff
@@ -1,11 +1,12 @@
-run: build
-	docker-compose up
+run:
+	docker-compose up --build
```
Comment (mattgen88) [inline Makefile]: Prod-like environment

File: Makefile [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
 build:
 	docker build -t exlibris:latest .
 run-local:
-	docker-compose down
-	docker-compose -f docker-compose.local.yml up
+	docker-compose -f docker-compose.local.yml down
+	docker-compose -f docker-compose.local.yml up -d
+	npm run build
 	go run main.go
```
Comment (mattgen88) [inline Makefile]: Runs the node build for vue and runs the main.go file locally, dependencies in docker

File: Makefile [hunk 5]
```diff
@@ -1,11 +1,12 @@
...
 dev:
-	docker-compose -f docker-compose.local.yml down
-	docker-compose -f docker-compose.local.yml up --build
\ No newline at end of file
+	docker-compose -f docker-compose.dev.yml down
+	docker-compose -f docker-compose.dev.yml up --build
```
Comment (mattgen88) [inline Makefile]: Full dev environment, with hot reload for api and app

File: main.go [hunk 6]
```diff
@@ -47,22 +46,25 @@ func main() {
...
+	api.HandleFunc("/register", h.Register).Methods(http.MethodPost, http.MethodOptions)
+	api.HandleFunc("/authenticate", h.Authenticate).Methods(http.MethodPost, http.MethodOptions)
+	api.HandleFunc("/book", h.SearchBooks)
+	api.HandleFunc("/book/{book}/read", h.Read).Methods(http.MethodPost, http.MethodOptions)
+	api.HandleFunc("/book/read", h.GetReads)
+	api.HandleFunc("/user/{username}/inbox", h.HandleInbox)
+	api.HandleFunc("/user/{username}/outbox", h.HandleOutbox)
+	api.HandleFunc("/@{username}/inbox", h.HandleInbox)
+	api.HandleFunc("/@{username}/outbox", h.HandleOutbox)
+	api.HandleFunc("/fedtest", h.FederationTest).Methods(http.MethodPost, http.MethodOptions)
```
Comment (mattgen88) [inline main.go]: API subrouter to make things cleaner

File: main.go [hunk 7]
```diff
@@ -47,22 +46,25 @@ func main() {
...
+
+	// App
+	r.PathPrefix("/").Handler(http.FileServer(http.Dir("./dist/")))
+	corsRouter := handlers.CORS(handlers.AllowedOrigins([]string{"*"}),
+		handlers.AllowedHeaders([]string{"Content-Type"}))
```
Comment (mattgen88) [inline main.go]: All important CORS...

File: src/Views/Home.vue [hunk 8]
```diff
@@ -26,13 +26,13 @@ export default {
   data() {
     return { books: [] };
   },
-  methods: { 	
+  methods: {
     onTermChange(searchTerm) {
-      axios.get('https://exlibris-fed.herokuapp.com/book', {
+      axios.get(process.env.VUE_APP_API_ORIGIN+'/book', {
```
Comment (mattgen88) [inline src/Views/Home.vue]: VUE_APP_* are compiled into the build

**Comment by mattgen88** [comment]
ping @Sigafoos

**Comment by Sigafoos** [comment]
What's the build/deploy process look like after this? Do we get rid of Netlify?

**Pull request merged**